Server/Hubs/CircuitConnection.cs

```diff
@@ -440,5 +444,64 @@ private void ProcessMessages()
                 }
             }
         }
+
+        public Task WOL(string deviceId)
+        {
+            var device = _dataService.GetDevice(deviceId);
+            if (device == null)
+            {
+                return Task.CompletedTask;
+            }
+
+            if (!_dataService.DoesUserHaveAccessToDevice(deviceId, _dataService.GetUserByNameWithOrg(User.UserName)))
+            {
+                return Task.CompletedTask;
+            }
+
+            var devices = _dataService.GetDevicesForUser(User.UserName);
+            var ActiveDevices = devices.Where(c => c.DeviceGroup == device.DeviceGroup && c.OrganizationID == device.OrganizationID).Select(s => s.ID).ToArray();
+
+            var connections = GetActiveClientConnections(ActiveDevices);
+            foreach (var connection in connections)
+            {
+                string[] MACs = device.MACAddresses.Split(new char[] { ',' }, StringSplitOptions.RemoveEmptyEntries);
+                foreach (var mac in MACs) {
+                    _agentHubContext.Clients.Client(connection.Key).SendAsync("WOL", mac);
+                }
+
+                break;  // Delete the line so that all active devices will send the Magic packet. 
+            }
+            
+            return Task.CompletedTask;
+        }
+
+        public Task WOL_ALL()
+        {
+            var orgdevices = _dataService.GetDevicesForUser(User.UserName);
+
+            var allAgenthubDevices = AgentHub.ServiceConnections.Where(x => x.Value.OrganizationID == User.OrganizationID).Select(c => c.Value.ID).ToArray();
+
+            var notActiveDevices = orgdevices.Where(c => !allAgenthubDevices.Contains(c.ID)).ToList();
+
+            foreach (var orgdevice in notActiveDevices) {
+
+                var activeDevices = AgentHub.ServiceConnections.Where(x =>
+                        x.Value.OrganizationID == User.OrganizationID
+                        && x.Value.DeviceGroupID == orgdevice.DeviceGroupID)
+                        .ToList();
+
+                foreach (var connection in activeDevices)
```

Comment:
We're enumerating collections here more than necessary (multiplicative time cost), resulting in unnecessary CPU use.

I'd recommend enumerating the connections once and arranging them into two dictionaries: one by device ID and one by device group ID.

Within the loop(s), these dictionaries can be used to retrieve devices and device groups with a constant time cost.